elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchAnalysisDefinitionProviderIT.java

```diff
@@ -0,0 +1,342 @@
+import org.hibernate.search.annotations.Indexed;
+import org.hibernate.search.annotations.TokenizerDef;
+import org.hibernate.search.backend.spi.Work;
+import org.hibernate.search.backend.spi.WorkType;
+import org.hibernate.search.elasticsearch.analyzer.definition.ElasticsearchAnalysisDefinitionRegistryBuilder;
+import org.hibernate.search.elasticsearch.analyzer.definition.spi.ElasticsearchAnalysisDefinitionProvider;
+import org.hibernate.search.elasticsearch.cfg.ElasticsearchEnvironment;
+import org.hibernate.search.elasticsearch.spi.ElasticsearchIndexManagerType;
+import org.hibernate.search.engine.integration.impl.ExtendedSearchIntegrator;
+import org.hibernate.search.exception.SearchException;
+import org.hibernate.search.query.engine.spi.EntityInfo;
+import org.hibernate.search.query.engine.spi.HSQuery;
+import org.hibernate.search.spi.SearchIntegrator;
+import org.hibernate.search.spi.SearchIntegratorBuilder;
+import org.hibernate.search.testsupport.TestForIssue;
+import org.hibernate.search.testsupport.setup.SearchConfigurationForTest;
+import org.hibernate.search.testsupport.setup.TransactionContextForTest;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+/**
+ * @author Yoann Rodiere
+ */
+@TestForIssue(jiraKey = "HSEARCH-2418")
+public class ElasticsearchAnalysisDefinitionProviderIT {
+
+	private static final String CUSTOM_ANALYZER_NAME = "custom-analyzer";
+
+	private static final String CUSTOM_ANALYZER_2_NAME = "custom-analyzer-2";
+
+	@Rule
+	public ExpectedException thrown = ExpectedException.none();
+
+	@Test
+	public void simple() {
+		ExtendedSearchIntegrator integrator = init( CustomAnalyzerProvider.class, CustomAnalyzerEntity.class );
+
+		assertThat( integrator.getAnalyzerRegistry( ElasticsearchIndexManagerType.INSTANCE )
+						.getAnalyzerReference( CUSTOM_ANALYZER_NAME ) )
+				.as( "Analyzer reference for '" + CUSTOM_ANALYZER_NAME + "' fetched from the integrator" )
+				.isNotNull();
+
+		CustomAnalyzerEntity entity = new CustomAnalyzerEntity();
+		entity.id = 0;
+		entity.field = "charFilterShouldReplace|foo";
+		index( integrator, entity );
+		assertMatchesExactly( integrator, entity, "field", "charfilterdidreplace" );
+	}
+
+	@Test
+	public void override() {
+		ExtendedSearchIntegrator integrator = init( CustomAnalyzerProvider.class, AnalyzerDefAnnotationEntity.class );
+
+		assertThat( integrator.getAnalyzerRegistry( ElasticsearchIndexManagerType.INSTANCE )
+						.getAnalyzerReference( CUSTOM_ANALYZER_NAME ) )
+				.as( "Analyzer reference for '" + CUSTOM_ANALYZER_NAME + "' fetched from the integrator" )
+				.isNotNull();
+
+		AnalyzerDefAnnotationEntity entity = new AnalyzerDefAnnotationEntity();
+		entity.id = 0;
+		entity.field = "charFilterShouldReplace|foo";
+		index( integrator, entity );
+		// Expecting a keyword tokenizer (as per the @AnalyzerDef annotation)
+		assertMatchesExactly( integrator, entity, "field", "charFilterShouldReplace|foo" );
```

Comment:
BTW the comment on line 80 mentioned "fetched from the integrator" so that's not consistent.